File: py/password_presets/generate.py
"""Core password generation logic using Python's secrets module."""

import secrets
from .wordlist import EFF_LARGE


def _secure_choice(seq: str | list) -> str:
    """Return a cryptographically secure random element from a sequence."""
    return secrets.choice(seq)


def generate_char(length: int, pools: tuple[str, ...]) -> str:
    """
    Generates a single character-based password.

    Guarantees at least one character from each pool, fills the remainder
    from the combined pool, then shuffles using Fisher-Yates via secrets.

    Args:
        length: Total password length.
        pools:  Tuple of character pool strings.

    Returns:
        A password string of the given length.
    """
    all_chars = "".join(pools)
    result: list[str] = []

    # Guarantee at least one character per pool
    for pool in pools:
        result.append(_secure_choice(pool))

    # Fill remaining positions from combined pool
    while len(result) < length:
        result.append(_secure_choice(all_chars))

    # Fisher-Yates shuffle using secrets.randbelow (no modulo bias)
    for i in range(len(result) - 1, 0, -1):
        j = secrets.randbelow(i + 1)
        result[i], result[j] = result[j], result[i]

    return "".join(result)


def generate_passphrase(word_count: int, separator: str) -> str:
    """
    Generates a single EFF-wordlist passphrase.

    Args:
        word_count: Number of words.
        separator:  String placed between words.

    Returns:
        A passphrase string.
    """
    words = [_secure_choice(EFF_LARGE) for _ in range(word_count)]
    return separator.join(words)

File: js/generate.js
import { secureRandomIndex, shuffle } from "./crypto.js";
import { EFF_LARGE } from "./wordlist.js";

/**
 * Generates a single character-based password from pool definitions.
 * Guarantees at least one character from each pool, then fills the
 * remainder from the combined pool, then shuffles.
 *
 * @param {number} length
 * @param {string[]} pools
 * @returns {string}
 */
export function generateChar(length, pools) {
  const allChars = pools.join("");
  const result = [];

  // Guarantee at least one character per pool
  for (const pool of pools) {
    result.push(pool[secureRandomIndex(pool.length)]);
  }

  // Fill remaining positions from combined pool
  while (result.length < length) {
    result.push(allChars[secureRandomIndex(allChars.length)]);
  }

  // Shuffle so guaranteed positions are not predictable
  return shuffle(result).join("");
}

/**
 * Generates a single EFF-wordlist passphrase.
 *
 * @param {number} wordCount
 * @param {string} separator
 * @returns {string}
 */
export function generatePassphrase(wordCount, separator) {
  const words = Array.from(
    { length: wordCount },
    () => EFF_LARGE[secureRandomIndex(EFF_LARGE.length)]
  );
  return words.join(separator);
}
